xdsl/builder.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+
+from typing import Callable, TypeAlias, overload
+
+from dataclasses import dataclass
+
+from xdsl.ir import OperationInvT, Attribute, Region, Block, BlockArgument
+from xdsl.dialects.builtin import FunctionType
+
+
+@dataclass
+class OpBuilder:
+    """
+    A helper class to construct IRs, by keeping track of where to insert an operation.
+    Currently can only append an operation to a given block, in the future will mirror the
+    API of `OpBuilder` in MLIR.
+
+    https://mlir.llvm.org/doxygen/classmlir_1_1OpBuilder.html
+    """
+
+    block: Block
+    """
+    Operations will be inserted in this block.
+    """
+
+    def insert(self, op: OperationInvT) -> OperationInvT:
+        """
+        Inserts `op` in `self.block` at the current insertion point.
+        """
+
+        self.block.add_op(op)
+        return op
+
+    @staticmethod
+    def region(func: Callable[[OpBuilder], None]) -> Region:
+        """
+        Generates a region given a function.
+        """
+
+        block = Block()
+        builder = OpBuilder(block)
+
+        func(builder)
+
+        return Region.from_block_list([block])
+
+    @staticmethod
+    def _callable_region_args(
+        types: tuple[list[Attribute], list[Attribute]]
+    ) -> Callable[[_CallableRegionFuncType], tuple[Region, FunctionType]]:
+        """
+        Constructs a tuple of (Region, FunctionType) for a region that takes some
+        arguments, and may return some results. The types of the arguments and results
```

Comment:
Regions should not return anything right?

Reply:
The `Region` doesn't, but the function that the region implements does. It feels weird to separate the input types from the function type, and I couldn't find a nicer API for how to pass them in. My goal was to minimise room for errors, like mismatch in function type and block argument types. With the API in this PR, that's close to impossible to get wrong. If we separate the return types from the input types, I can imagine it being much easier.